nav2_map_server/src/map_saver.cpp

```diff
@@ -1,110 +1,203 @@
+    } catch (std::invalid_argument &) {
+      map_mode = MapMode::Trinary;
+      RCLCPP_WARN(
+        get_logger(), "Map mode parameter not recognized: '%s', using default value (trinary)",
+        mode_str.c_str());
+    }
 
-using namespace std::chrono_literals;
+    image_format = declare_parameter("image_format", map_mode == MapMode::Scale ? "png" : "pgm");
 
-int main(int argc, char ** argv)
+    RCLCPP_INFO(get_logger(), "Waiting for the map");
+    map_sub_ = create_subscription<nav_msgs::msg::OccupancyGrid>(
+      "map", rclcpp::SystemDefaultsQoS(),
+      std::bind(&MapSaver::mapCallback, this, std::placeholders::_1));
+  }
+}
+
+void MapSaver::try_write_map_to_file(const nav_msgs::msg::OccupancyGrid & map)
 {
-  rclcpp::init(argc, argv);
-  rclcpp::Logger logger = rclcpp::get_logger("map_saver");
-
-  std::string mapname = "map";
-  int threshold_occupied = 65;
-  int threshold_free = 25;
-
-  for (int i = 1; i < argc; i++) {
-    if (!strcmp(argv[i], "-h")) {
-      puts(USAGE);
-      return 0;
-    } else if (!strcmp(argv[i], "-f")) {
-      if (++i < argc) {
-        mapname = argv[i];
-      } else {
-        puts(USAGE);
-        return 1;
-      }
-    } else if (!strcmp(argv[i], "--occ")) {
-      if (++i < argc) {
-        threshold_occupied = atoi(argv[i]);
-        if (threshold_occupied < 1 || threshold_occupied > 100) {
-          RCLCPP_ERROR(logger, "Threshold_occupied must be between 1 and 100");
-          return 1;
-        }
-      } else {
-        puts(USAGE);
-        return 1;
-      }
-    } else if (!strcmp(argv[i], "--free")) {
-      if (++i < argc) {
-        threshold_free = atoi(argv[i]);
-        if (threshold_free < 0 || threshold_free > 100) {
-          RCLCPP_ERROR(logger, "Threshold_free must be between 0 and 100");
-          return 1;
+  rclcpp::Logger logger = this->get_logger();
+  RCLCPP_INFO(
+    logger, "Received a %d X %d map @ %.3f m/pix", map.info.width, map.info.height,
+    map.info.resolution);
+
+  Magick::CoderInfo info(image_format);
+  if (!info.isWritable()) {
+    image_format = "png";
+    RCLCPP_WARN(logger, "Format is not writable. Falling back to %s", image_format.c_str());
+  }
+
+  std::string mapdatafile = mapname_ + "." + image_format;
+  {
+    // should never see this color, so the initialization value is just for debugging
+    Magick::Image image({map.info.width, map.info.height}, "red");
+
+    // In scale mode, we need the alpha (matte) channel. Else, we don't.
+    // NOTE: GraphicsMagick seems to have trouble loading the alpha channel when saved with
+    // Magick::GreyscaleMatte, so we use TrueColorMatte instead.
+    image.type(map_mode == MapMode::Scale ? Magick::TrueColorMatteType : Magick::GrayscaleType);
+
+    // Since we only need to support 100 different pixel levels, 8 bits is fine
+    image.depth(8);
+
+    if (
+      map_mode == MapMode::Scale &&
+      (image_format == "pgm" || image_format == "jpg" || image_format == "jpeg"))
```

Comment:
I'm not sure I totally understand why we're supporting multiple filetypes. Shouldn't we want to keep this consistent so other applications that are using these maps don't need to change? Is the size of the pgm large enough over your cellular connection the compression is important?

Reply:
pgm does not support transparency. We need transparency for scale mode.
Also, I'd like a format with good tooling support. Browsers tend to support png and tend not to support pgm.

Reply:
I thought we figured out that we could do scale with pgm but it was just the saver that didn't allow it? I definitely have scaled pgm files in ROS navigation1.

Reply:
No, you can't. At least not as documented.

Reply:
I unfortunately cannot point you to code as this was a concept from a prior proprietary project, but here's a thread [discussing it](https://answers.ros.org/question/294462/nav-stack-one-map-for-amcl-and-one-for-move_base/)

Perhaps I'm missing something, but I believe this is what we're talking about

Reply:
Did you send me the right thread? I can't see mention of trying to use map_mode = scale.

You *can* do it with map_mode = raw, which is not a great editing experience. Just like our previous use of SDL in map loader which supported any image format known to SDL, I don't see a need to place arbitrary restrictions on the savable formats. If it turns out to be a problem or the library handles some formats poorly, we can add guardrails later.

I do think that PNGs are nice, so whatever restrictions we add, I'd like to ensure we keep support for them. They are losslessly compressed and supported well by tools. For instance, if you check a [png into your repo](https://github.com/ros-planning/navigation2/blob/master/nav2_map_server/test/testmap.png), you can view it on GitHub. Not as easily [for bmp](https://github.com/ros-planning/navigation2/blob/master/nav2_map_server/test/testmap.bmp) and not at all [for pgm](https://github.com/ros-planning/navigation2/blob/master/nav2_costmap_2d/test/map/TenByTen.pgm).

Reply:
Got it, ok that's where we've missed each other - makes sense. 

I agree pngs are useful. I'd be in favor of restricting to pgm & png so we dont have to keep adding more and more. Frankly I'd also be in favor of just nuking pgm support entirely if we also provided a script to convert old pgm files to png (or just instructions/link if we dont want to write it)